power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/linear_pf_solver.hpp

```diff
@@ -123,6 +93,30 @@ template <bool sym> class LinearPFSolver {
     SparseLUSolver<ComplexTensor<sym>, ComplexValue<sym>, ComplexValue<sym>> sparse_solver_;
     typename SparseLUSolver<ComplexTensor<sym>, ComplexValue<sym>, ComplexValue<sym>>::BlockPermArray perm_;
 
+    void prepare_matrix_and_rhs(YBus<sym> const& y_bus, PowerFlowInput<sym> const& input, MathOutput<sym>& output) {
+        // getter
+        IdxVector const& bus_entry = y_bus.lu_diag();
+        IdxVector const& load_gen_bus_idxptr = *load_gen_bus_indptr_;
+        IdxVector const& source_bus_indptr = *source_bus_indptr_;
+        for (Idx bus_number = 0; bus_number != n_bus_; ++bus_number) {
+            Idx const diagonal_position = bus_entry[bus_number];
+            auto& diagonal_element = mat_data_[diagonal_position];
+            auto& u_bus = output.u[bus_number];
+            add_loads(load_gen_bus_idxptr, bus_number, input, diagonal_element);
+            shared_solver_functions::add_sources<sym>(source_bus_indptr, bus_number, y_bus, input.source,
```

Comment:
i don't know about you but i really like this explicit way of writing functions that tells us which part is solver-specific and which part is "normal" and "obvious"